c/extras/ftrl.cc

```diff
@@ -0,0 +1,527 @@
+                             auto d_int = c_int->elements_r();
+                             index = static_cast<uint64_t>(d_int[row_id]);
+                             break;
+                           }
+
+      // For doubles, use their bit representation as `uint64_t`.
+      case LType::REAL:    {
+                             auto c_real = static_cast<RealColumn<double>*>(c);
+                             auto d_real = c_real->elements_r();
+                             index = hash_double(d_real[row_id]);
+                             break;
+                           }
+
+      // For strings, invoke `hash_string()` that will call the actual
+      // hash function based on the provided `hash_type`.
+      case LType::STRING:  {
+                             auto c_string = static_cast<StringColumn<uint32_t>*>(c);
+                             auto d_string = c_string->offsets();
+                             const char* strdata = c_string->strdata();
+                             const uint32_t strstart = d_string[row_id-1] & ~GETNA<uint32_t>();
+                             const char* c_str = strdata + strstart;
+                             uint32_t len = d_string[row_id] - strstart;
+                             index = hash_string(c_str, len * sizeof(char));
+                             break;
+                           }
+      default:             throw ValueError() << "Datatype is not supported!";
+    }
+    // Add the column name hash to the hashed value, so that the same value
+    // in different columns will result in different hashes.
+    uint64_t h = hash_string(c_names[i].c_str(), c_names[i].length() * sizeof(char));
+    index += h;
+    x[i+1] = index % d;
+  }
+
+  // Do feature interaction if required. We may also want to test
+  // just a simple `h = x[i+1] + x[j+1]` approach.
+  size_t count = 0;
+  if (inter) {
+    for (size_t i = 0; i < n_features - 1; ++i) {
+      for (size_t j = i + 1; j < n_features - 1; ++j) {
+        std::string s = std::to_string(x[i+1]) + std::to_string(x[j+1]);
+        uint64_t h = hash_string(s.c_str(), s.length() * sizeof(char));
+        x[n_features + count] = h % d;
+        count++;
+      }
+    }
+  }
+}
+
+
+/*
+*  Calculate logloss based on a prediction and the actual target.
+*/
+double Ftrl::logloss(double p, bool target) {
+  double epsilon = std::numeric_limits<double>::epsilon();
+  p = std::max(std::min(p, 1 - epsilon), epsilon);
+  if (target) {
+    return -log(p);
+  } else {
+    return -log(1 - p);
+  }
+}
+
+
+/*
+*  Calculate signum.
+*/
+inline double Ftrl::signum(double x) {
+  if (x > 0) return 1;
+  if (x < 0) return -1;
+  return 0;
+}
+
+
+/*
+*  Hash `double` to `uint64_t` based on its bit representation.
+*/
+inline uint64_t Ftrl::hash_double(double x) {
+  uint64_t* h = reinterpret_cast<uint64_t*>(&x);
+  return *h;
+}
+
+
+/*
+*  Murmur2 hash function by Austin Appleby.
+*  More details at https://sites.google.com/site/murmurhash/
```

Comment:
If some code is taken from external sources, the best practice is to:
1) Put that code in a separate file
2) Clearly note where it was taken from, and what was the license
3) Make sure the conditions of the license are satisfied.
For example, the link above says the code is licensed MIT. However, MIT license requires that "The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software."

Reply:
I'm not sure about the MIT license. This is what you get when go to the site where this code moved at some point

//-----------------------------------------------------------------------------
// MurmurHash3 was written by Austin Appleby, and is placed in the public
// domain. The author hereby disclaims copyright to this source code.

https://github.com/aappleby/smhasher/blob/master/src/MurmurHash2.cpp
https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp

What do you think we should do?

Reply:
Actually I just noticed that the link says "Update March 1, 2011 - The 32-bit variant of MurmurHash3 is finalized, and this page is deprecated", so it's probably better not to use it and replace with https://github.com/aappleby/smhasher On the latter It says "This is the home for the MurmurHash family of hash functions along with the SMHasher test suite used to verify them. SMHasher is released under the MIT license. All MurmurHash versions are public domain software, and the author disclaims all copyright to their code." Do you think it means `MurmurHash` is licensed under MIT or not?

Reply:
It's not.
In the copyright law, the owner (author) of the code has an exclusive right to decide what other people can or cannot do to their code. That decision is embodied in the form of a LICENSE. For example, MIT license says "do pretty much anything, but don't remove this notice".

However, once the author puts his code into the public domain, he is no longer the owner and cannot grant licenses. The code can be used by anyone, without any pre-conditions. 

Thus, the best approach would be to: (1) put it into a separate file, (2) add a notice at the top of the file that this code is written by Austin Appleby and placed into public domain, (3) add a URL where it was borrowed from.
Do not add our standard MIT license boilerplate, because that would require us to claim the copyright in the work, which is technically possible but not very fair.